pulsar-flink-1.11/src/main/java/org/apache/flink/table/PulsarDynamicTableSource.java

```diff
@@ -0,0 +1,237 @@
+                                    String serviceUrl,
+                                    String adminUrl,
+                                    Properties properties,
+                                    PulsarOptions.StartupOptions startupOptions) {
+        this.outputDataType = outputDataType;
+        this.decodingFormat = decodingFormat;
+        this.topic = topic;
+        this.serviceUrl = serviceUrl;
+        this.adminUrl = adminUrl;
+        properties.computeIfAbsent("topic", (k) -> topic);
+        this.properties = properties;
+        this.startupOptions = startupOptions;
+    }
+
+    @Override
+    public ChangelogMode getChangelogMode() {
+        return ChangelogMode.insertOnly();
+    }
+
+    @Override
+    public ScanRuntimeProvider getScanRuntimeProvider(ScanContext runtimeProviderContext) {
+
+        DeserializationSchema<RowData> deserializationSchema =
+                this.decodingFormat.createRuntimeDecoder(runtimeProviderContext, this.outputDataType);
+
+        FlinkPulsarSource<RowData> source = new FlinkPulsarSource<>(
+                adminUrl,
+                newClientConf(serviceUrl),
+                new PulsarDeserializationSchemaWrapper<>(deserializationSchema),
+                properties
+        );
+        // TODO 调整结构
+        switch (startupOptions.startupMode) {
+            case EARLIEST:
+                source.setStartFromEarliest();
+                break;
+            case LATEST:
+                source.setStartFromLatest();
+                break;
+            case SPECIFIC_OFFSETS:
+                source.setStartFromSpecificOffsets(startupOptions.specificOffsets);
+                break;
+            case EXTERNAL_SUBSCRIPTION:
+                MessageId subscriptionPosition = MessageId.latest;
+                if (CONNECTOR_STARTUP_MODE_VALUE_EARLIEST
+                        .equals(properties.get(CONNECTOR_EXTERNAL_SUB_DEFAULT_OFFSET))) {
+                    subscriptionPosition = MessageId.earliest;
+                }
+                source.setStartFromSubscription(startupOptions.externalSubscriptionName, subscriptionPosition);
+        }
+        return SourceFunctionProvider.of(source, false);
+    }
+
+    @Override
+    public DynamicTableSource copy() {
+        return new PulsarDynamicTableSource(
+                this.outputDataType,
+                this.decodingFormat,
+                this.topic,
+                this.serviceUrl,
+                this.adminUrl,
+                this.properties,
+                this.startupOptions
+        );
+    }
+
+    @Override
+    public String asSummaryString() {
+        return "Pulsar universal table source";
+    }
+
+    private static ClientConfigurationData newClientConf(String serviceUrl) {
+        ClientConfigurationData clientConf = new ClientConfigurationData();
+        clientConf.setServiceUrl(serviceUrl);
+        return clientConf;
+    }
+
+    private static Properties removeConnectorPrefix(Properties in) {
```

Comment:
This method seems to be repeated to PulsarDynamicTableFactory.class . Can we extract it and put it into the option class or some utils class.